File: bindings/php/src/Rendering.php
<?php

declare(strict_types=1);

namespace Sone;

/** Page margins. A single number applies to all four sides. */
final class Margin implements \JsonSerializable
{
    public function __construct(
        public readonly int|float $top = 0,
        public readonly int|float $right = 0,
        public readonly int|float $bottom = 0,
        public readonly int|float $left = 0,
    ) {}

    public static function all(int|float $value): self
    {
        return new self($value, $value, $value, $value);
    }

    public function jsonSerialize(): array
    {
        return ['top' => $this->top, 'right' => $this->right, 'bottom' => $this->bottom, 'left' => $this->left];
    }
}

/** A font the document carries with it, so another sone engine renders it identically. */
final class FontSource
{
    public function __construct(public readonly string $name, public readonly string $src) {}
}

/** A node plus its render configuration, with one method per output format. */
final class Rendering
{
    private const FORMAT_BY_EXTENSION = [
        'png' => OutputFormat::Png, 'jpg' => OutputFormat::Jpeg, 'jpeg' => OutputFormat::Jpeg,
        'webp' => OutputFormat::Webp, 'pdf' => OutputFormat::Pdf, 'svg' => OutputFormat::Svg,
        'raw' => OutputFormat::Raw, 'rgba' => OutputFormat::Raw,
    ];

    private ?string $cached = null;

    /** @param list<FontSource> $fonts */
    public function __construct(
        private readonly Node $root,
        private readonly ?Engine $engine = null,
        private readonly int|float|null $width = null,
        private readonly int|float|null $height = null,
        private readonly ?string $background = null,
        private readonly int|float|null $density = null,
        private readonly int|float|null $pageHeight = null,
        private readonly Margin|int|float|null $margin = null,
        private readonly LastPageHeight|string|null $lastPageHeight = null,
        private readonly ?Node $header = null,
        private readonly ?Node $footer = null,
        private readonly array $fonts = [],
    ) {}

    public function engine(): Engine
    {
        return $this->engine ?? Engine::default();
    }

    // ── the document ────────────────────────────────────────────────────────

    /** The IR document as an array. */
    public function toIr(): array
    {
        $config = array_filter([
            'width' => $this->width,
            'height' => $this->height,
            'background' => $this->background,
            'density' => $this->density,
            'pageHeight' => $this->pageHeight,
            'margin' => \is_object($this->margin) ? $this->margin->jsonSerialize() : $this->margin,
            'lastPageHeight' => $this->lastPageHeight instanceof LastPageHeight
                ? $this->lastPageHeight->value
                : $this->lastPageHeight,
            'header' => $this->header?->toIr(),
            'footer' => $this->footer?->toIr(),
        ], static fn (mixed $value): bool => $value !== null);

        $document = ['sone' => 1];
        if ($this->fonts !== []) {
            $document['fonts'] = array_map(
                static fn (FontSource $font): array => ['name' => $font->name, 'src' => $font->src],
                $this->fonts,
            );
        }
        if ($config !== []) {
            $document['config'] = $config;
        }
        $document['root'] = $this->root->toIr();

        return $document;
    }

    /** The IR document as JSON, built once and reused. */
    public function toJson(bool $pretty = false): string
    {
        $flags = JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES;
        if ($pretty) {
            return json_encode($this->toIr(), $flags | JSON_PRETTY_PRINT);
        }

        return $this->cached ??= json_encode($this->toIr(), $flags);
    }

    // ── outputs ─────────────────────────────────────────────────────────────

    public function png(?float $density = null): string
    {
        return $this->engine()->render($this->toJson(), OutputFormat::Png, $density);
    }

    public function jpeg(float $quality = 1.0, ?float $density = null): string
    {
        return $this->engine()->render($this->toJson(), OutputFormat::Jpeg, $density, $quality);
    }

    public function webp(float $quality = 1.0, ?float $density = null): string
    {
        return $this->engine()->render($this->toJson(), OutputFormat::Webp, $density, $quality);
    }

    /** Raw RGBA pixels, row-major, unpremultiplied. */
    public function raw(?float $density = null): string
    {
        return $this->engine()->render($this->toJson(), OutputFormat::Raw, $density);
    }

    /** A PDF. With `pageHeight` set, one page per break and selectable text. */
    public function pdf(): string
    {
        return $this->engine()->render($this->toJson(), OutputFormat::Pdf);
    }

    public function svg(): string
    {
        return $this->engine()->render($this->toJson(), OutputFormat::Svg);
    }

    /** One raster image per page. Requires `pageHeight`. @return list<string> */
    public function pages(OutputFormat|string $format = OutputFormat::Png, ?float $density = null, float $quality = 1.0): array
    {
        return $this->engine()->renderPages($this->toJson(), $format, $density, $quality);
    }

    /** Render and write to `$path`, inferring the format from its extension. */
    public function save(string $path, ?float $density = null, float $quality = 1.0): string
    {
        $bytes = match ($this->formatFor($path)) {
            OutputFormat::Png => $this->png($density),
            OutputFormat::Jpeg => $this->jpeg($quality, $density),
            OutputFormat::Webp => $this->webp($quality, $density),
            OutputFormat::Raw => $this->raw($density),
            OutputFormat::Pdf => $this->pdf(),
            OutputFormat::Svg => $this->svg(),
        };
        file_put_contents($path, $bytes);

        return $path;
    }

    /** Write `name-1.png`, `name-2.png`, … next to `$path`. @return list<string> */
    public function savePages(string $path, ?float $density = null, float $quality = 1.0): array
    {
        $extension = pathinfo($path, PATHINFO_EXTENSION);
        $stem = $extension === '' ? $path : substr($path, 0, -\strlen($extension) - 1);
        $suffix = $extension === '' ? '.png' : ".$extension";
        $format = self::FORMAT_BY_EXTENSION[strtolower($extension)] ?? OutputFormat::Png;

        $written = [];
        foreach ($this->pages($format, $density, $quality) as $index => $bytes) {
            $name = $stem . '-' . ($index + 1) . $suffix;
            file_put_contents($name, $bytes);
            $written[] = $name;
        }

        return $written;
    }

    // ── introspection ───────────────────────────────────────────────────────

    /** The computed layout tree. */
    public function layout(): array
    {
        return json_decode($this->engine()->dumpLayout($this->toJson()), true);
    }

    /** Dataset-style boxes at node, line or word granularity. */
    public function metadata(Granularity|string $granularity = Granularity::Node): array
    {
        return json_decode($this->engine()->dumpMetadata($this->toJson(), $granularity), true);
    }

    private function formatFor(string $path): OutputFormat
    {
        $extension = strtolower(pathinfo($path, PATHINFO_EXTENSION));

        return self::FORMAT_BY_EXTENSION[$extension]
            ?? throw new \InvalidArgumentException("cannot infer an output format from \"$path\"");
    }
}

/**
 * Wrap a node with render configuration.
 *
 *     Sone::render($root, density: 2)->save('card.png');
 */
final class Sone
{
    /** @param list<FontSource> $fonts */
    public static function render(
        Node $root,
        ?Engine $engine = null,
        int|float|null $width = null,
        int|float|null $height = null,
        ?string $background = null,
        int|float|null $density = null,
        int|float|null $pageHeight = null,
        Margin|int|float|null $margin = null,
        LastPageHeight|string|null $lastPageHeight = null,
        ?Node $header = null,
        ?Node $footer = null,
        array $fonts = [],
    ): Rendering {
        return new Rendering($root, $engine, $width, $height, $background, $density,
            $pageHeight, $margin, $lastPageHeight, $header, $footer, $fonts);
    }
}

File: bindings/dart/lib/src/rendering.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'engine.dart';
import 'keywords.dart';
import 'native.dart';
import 'node.dart';

/// Page margins. A single number applies to all four sides.
class Margin {
  const Margin({this.top = 0, this.right = 0, this.bottom = 0, this.left = 0});

  const Margin.all(num value)
      : top = value,
        right = value,
        bottom = value,
        left = value;

  final num top;
  final num right;
  final num bottom;
  final num left;

  Map<String, Object?> toIr() =>
      {'top': top, 'right': right, 'bottom': bottom, 'left': left};
}

/// A font the document carries with it, so another sone engine — the CLI, say —
/// renders it identically.
class FontSource {
  const FontSource(this.name, this.src);

  final String name;
  final String src;
}

/// A node plus its render configuration, with one method per output format.
class Rendering {
  Rendering(
    this.root, {
    this.engine,
    this.width,
    this.height,
    this.background,
    this.density,
    this.pageHeight,
    this.margin,
    this.lastPageHeight,
    this.header,
    this.footer,
    this.fonts,
  });

  final Node root;
  final Engine? engine;
  final num? width;
  final num? height;
  final String? background;
  final num? density;
  final num? pageHeight;
  final Margin? margin;
  final LastPageHeight? lastPageHeight;

  /// Drawn at the top of every page. Use the literal tokens `{pageNumber}` and
  /// `{totalPages}` — the engine substitutes them.
  final Node? header;

  /// Drawn at the bottom of every page.
  final Node? footer;

  final List<FontSource>? fonts;

  Engine get _engine => engine ?? Engine.instance;

  String? _cached;

  // ── the document ────────────────────────────────────────────────────────

  /// The IR document as a map.
  Map<String, Object?> toIr() {
    final config = <String, Object?>{
      if (width != null) 'width': width,
      if (height != null) 'height': height,
      if (background != null) 'background': background,
      if (density != null) 'density': density,
      if (pageHeight != null) 'pageHeight': pageHeight,
      if (margin != null) 'margin': margin!.toIr(),
      if (lastPageHeight != null) 'lastPageHeight': lastPageHeight!.value,
      if (header != null) 'header': header!.toIr(),
      if (footer != null) 'footer': footer!.toIr(),
    };
    return <String, Object?>{
      'sone': 1,
      if (fonts != null && fonts!.isNotEmpty)
        'fonts': [
          for (final font in fonts!) {'name': font.name, 'src': font.src},
        ],
      if (config.isNotEmpty) 'config': config,
      'root': root.toIr(),
    };
  }

  /// The IR document as JSON, built once and reused.
  String toJsonString({bool pretty = false}) {
    if (pretty) return const JsonEncoder.withIndent('  ').convert(toIr());
    return _cached ??= jsonEncode(toIr());
  }

  // ── outputs ─────────────────────────────────────────────────────────────

  Uint8List png({double? density}) => _engine.render(toJsonString(),
      format: OutputFormat.png, density: density);

  Uint8List jpeg({double quality = 1.0, double? density}) =>
      _engine.render(toJsonString(),
          format: OutputFormat.jpeg, density: density, quality: quality);

  Uint8List webp({double quality = 1.0, double? density}) =>
      _engine.render(toJsonString(),
          format: OutputFormat.webp, density: density, quality: quality);

  /// Raw RGBA pixels, row-major, unpremultiplied.
  Uint8List raw({double? density}) => _engine.render(toJsonString(),
      format: OutputFormat.raw, density: density);

  /// A PDF. With [pageHeight] set, one page per break and selectable text.
  Uint8List pdf() => _engine.render(toJsonString(), format: OutputFormat.pdf);

  Uint8List svg() => _engine.render(toJsonString(), format: OutputFormat.svg);

  /// One raster image per page. Requires [pageHeight].
  List<Uint8List> pages(
          {OutputFormat format = OutputFormat.png,
          double? density,
          double quality = 1.0}) =>
      _engine.renderPages(toJsonString(),
          format: format, density: density, quality: quality);

  /// Render and write to [path], inferring the format from its extension.
  File save(String path, {double? density, double quality = 1.0}) {
    final bytes = switch (_formatFor(path)) {
      OutputFormat.png => png(density: density),
      OutputFormat.jpeg => jpeg(quality: quality, density: density),
      OutputFormat.webp => webp(quality: quality, density: density),
      OutputFormat.raw => raw(density: density),
      OutputFormat.pdf => pdf(),
      OutputFormat.svg => svg(),
    };
    return File(path)..writeAsBytesSync(bytes);
  }

  /// Write `name-1.png`, `name-2.png`, … next to [path].
  List<File> savePages(String path, {double? density, double quality = 1.0}) {
    final dot = path.lastIndexOf('.');
    final stem = dot < 0 ? path : path.substring(0, dot);
    final extension = dot < 0 ? '.png' : path.substring(dot);
    final format = _formatFor(path, fallback: OutputFormat.png);
    final rendered = pages(format: format, density: density, quality: quality);
    return [
      for (var index = 0; index < rendered.length; index++)
        File('$stem-${index + 1}$extension')..writeAsBytesSync(rendered[index]),
    ];
  }

  // ── introspection ───────────────────────────────────────────────────────

  /// The computed layout tree.
  Map<String, Object?> layout() =>
      jsonDecode(_engine.dumpLayout(toJsonString())) as Map<String, Object?>;

  /// Dataset-style boxes at node, line or word granularity.
  Map<String, Object?> metadata([Granularity granularity = Granularity.node]) =>
      jsonDecode(_engine.dumpMetadata(toJsonString(), granularity.value))
          as Map<String, Object?>;

  static OutputFormat _formatFor(String path, {OutputFormat? fallback}) {
    final dot = path.lastIndexOf('.');
    final extension = dot < 0 ? '' : path.substring(dot).toLowerCase();
    return switch (extension) {
      '.png' => OutputFormat.png,
      '.jpg' || '.jpeg' => OutputFormat.jpeg,
      '.webp' => OutputFormat.webp,
      '.pdf' => OutputFormat.pdf,
      '.svg' => OutputFormat.svg,
      '.raw' || '.rgba' => OutputFormat.raw,
      _ => fallback ??
          (throw ArgumentError.value(
              path, 'path', 'cannot infer an output format')),
    };
  }
}

/// Wrap a node with render configuration.
///
///     render(root, density: 2).save('card.png');
Rendering render(
  Node root, {
  Engine? engine,
  num? width,
  num? height,
  String? background,
  num? density,
  num? pageHeight,
  Margin? margin,
  LastPageHeight? lastPageHeight,
  Node? header,
  Node? footer,
  List<FontSource>? fonts,
}) =>
    Rendering(root,
        engine: engine,
        width: width,
        height: height,
        background: background,
        density: density,
        pageHeight: pageHeight,
        margin: margin,
        lastPageHeight: lastPageHeight,
        header: header,
        footer: footer,
        fonts: fonts);
